### PR TITLE
refactor(protocol): Add fixed codec

### PR DESCRIPTION
**Pull request by mkxbl**
**What type of PR is this?**
refactor

**What this PR does / why we need it**:
add fixed codec to protocol

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by yejiayu** [comment]
Don't implement `FixedCodec` for `Tx Approve`, because we don't need it in the executor2.0 design.

File: protocol/src/fixed_codec/epoch.rs [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+
+        Ok(Proof {
+            epoch_id,
+            round,
+            epoch_hash,
+            signature,
+            bitmap
+        })
+    }
+}
```
Comment (KaoImin) [inline protocol/src/fixed_codec/epoch.rs]: The implement of `rlp::Decodable` may refer to this https://github.com/cryptape/overlord/blob/master/src/codec.rs#L19:L34

File: protocol/src/fixed_codec/tests/mod.rs [hunk 2]
```diff
@@ -0,0 +1,269 @@
+extern crate test;
+
+mod fixed_codec;
+
+use bytes::Bytes;
+use num_traits::FromPrimitive;
+use rand::random;
+use std::collections::BTreeMap;
```
Comment (KaoImin) [inline protocol/src/fixed_codec/tests/mod.rs]: Like this:
```rust
use std::collections::BTreeMap;

use bytes::Bytes;
use num_traits::FromPrimitive;
use rand::random;
```

**Comment by mkxbl** [comment]
> Don't implement `FixedCodec` for `Tx Approve`, because we don't need it in the executor2.0 design.

ok

**Comment by mkxbl** [comment]
> Needs `cargo fmt`

not finished yet

**Comment by yejiayu** [comment]
@zeroqn PTAL

File: protocol/src/fixed_codec/receipt.rs [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        })
+    }
+}
+
+const TRANSFER_RESULT_FLAG: u8 = 0;
+const DEPLOY_RESULT_FLAG: u8 = 1;
+const CALL_RESULT_FLAG: u8 = 2;
+const FAIL_RESULT_FLAG: u8 = 3;
+
+#[allow(unused_variables)]
```
Comment (yejiayu) [inline protocol/src/fixed_codec/receipt.rs]: Useless drive?

File: protocol/src/fixed_codec/receipt.rs [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+                    ContractType::Native => 3,
+                };
+                s.append(&type_flag);
+            }
+            ReceiptResult::Call {
+                contract,
+                return_value,
+                logs_bloom,
+            } => {
+                // TODO(@yejiayu): The interface for `call` is about to be modified.
```
Comment (yejiayu) [inline protocol/src/fixed_codec/receipt.rs]: Remove this TODO and implement it

File: protocol/src/fixed_codec/transaction.rs [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    primitive::{Balance, ContractType, Fee, Hash, UserAddress},
+    transaction::{CarryingAsset, RawTransaction, SignedTransaction, TransactionAction},
+};
+use crate::{impl_default_fixed_codec_for, ProtocolResult};
+
+// Impl ProtocolFixedCodec trait for types
+impl_default_fixed_codec_for!(transaction, [RawTransaction, SignedTransaction]);
+
+impl rlp::Encodable for RawTransaction {
+    fn rlp_append(&self, s: &mut rlp::RlpStream) {
```
Comment (yejiayu) [inline protocol/src/fixed_codec/transaction.rs]: Implementing so many nested `RawTransaction` encode can be confusing for the Toolchain.

Maybe we should flatten it? like this
````rs
s.append(TX_FLAG);
s.append(transfer_receiver);
s.append(transfer_amount);
s.append(fee_asset_id);
.....
````

Comment (mkxbl) [reply protocol/src/fixed_codec/transaction.rs]: ok. working on it

**Pull request merged**